### PR TITLE
detect/asn1: Migrate to Rust - v1

### PR DESCRIPTION
**Pull request by sharksforarms**
Draft PR: Request for review/comments

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/2977

Describe changes:
- Migration of ASN1 decoder and detect module to Rust

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

TODO:
- 4.0.0 release of https://github.com/rusticata/der-parser/
- https://github.com/rusticata/kerberos-parser/pull/3
- https://github.com/rusticata/snmp-parser/pull/7
- Once the above PRs are merged, update `rust/Cargo.toml.in`

### CODE REVIEW

**Comment by sharksforarms** [comment]
Blockers:

Rust 1.34.2 fails to compile der-parser and ipsec-parser

```rust
 Compiling der-parser v4.0.0-alpha3
error[E0658]: use of unstable library feature 'option_xor' (see issue #50512)
   --> /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/der-parser-4.0.0-alpha3/src/ber/ber.rs:392:33
    |
392 |                 if self.raw_tag.xor(other.raw_tag).is_none() {
    |                                 ^^^

   Compiling ipsec-parser v0.5.0
error[E0658]: use of unstable library feature 'iter_copied' (see issue #57127)
   --> /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/der-parser-4.0.0-alpha3/src/oid.rs:134:14
    |
134 |             .copied()
    |              ^^^^^^
```

**Comment by victorjulien** [comment]
cc @chifflier

File: rust/src/asn1/mod.rs [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+#[no_mangle]
+pub unsafe extern "C" fn rs_asn1_checks(ptr: *const Asn1, ad_ptr: *const DetectAsn1Data) -> u8 {
+    if ptr.is_null() || ad_ptr.is_null() {
+        return 0;
+    }
+
+    let asn1 = &*ptr;
+    let ad = &*ad_ptr;
+
+    match asn1.check(ad) {
```
Comment (victorjulien) [inline rust/src/asn1/mod.rs:302]: Could be written more compact:
```rust
match asn1.check(ad) {
￼        Ok(Some(_check)) => 1, // match
￼        Ok(None) => 0, // no match
```

File: rust/src/asn1/parse_rules.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+/// Struct to hold parsed asn1 keyword options
+#[derive(Debug, PartialEq)]
+pub struct DetectAsn1Data {
+    pub bitstring_overflow: bool,
+    pub double_overflow: bool,
+    pub oversize_length: Option<u32>,
+    pub absolute_offset: Option<u16>,
+    pub relative_offset: Option<i32>,
+    pub max_frames: usize,
```
Comment (victorjulien) [inline rust/src/asn1/parse_rules.rs:94]: Is `usize` like `size_t`, a type with a arch dependent size? I'd prefer an explicit type, like `u32` or `u64`. I *think* we parse this as an `u16`? So then we should use that.

File: rust/src/asn1/parse_rules.rs [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    Ok((rest, data))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_case::test_case;
+
+    // Test oversize_length
+    #[test_case("oversize_length 1024", DetectAsn1Data { oversize_length: Some(1024), ..Default::default()}; "check that we parse oversize_length correctly")]
```
Comment (victorjulien) [inline rust/src/asn1/parse_rules.rs:228]: can you update these to not exceed line length limit? ~80 char soft limit, 100 hard limit

File: rust/src/asn1/parse_rules.rs (hunk 2 above)
Comment (sharksforarms) [reply rust/src/asn1/parse_rules.rs:94]: No problem! For clarification, in general, should these types be avoided in suricata?

I use `usize` as a counter for recursion here: https://github.com/OISF/suricata/pull/4954/files#diff-d05f7cef7932b801ab5ff5266426ca11R91

Should this be changed?

File: rust/src/asn1/mod.rs [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        .checked_sub(offset)
+        .ok_or(Asn1DecodeError::InvalidKeywordParameter)?
+        .into();
+
+    // Apply offset to input pointer
+    let offset = isize::try_from(offset)?;
+    let input = unsafe { input.offset(offset) };
+
+    // Get the slice from memory
+    let slice = unsafe { std::slice::from_raw_parts(input, input_len) };
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply rust/src/asn1/mod.rs:241]: > the convention we use in the app parsers is to create the slice as early as possible in the unsafe FFI func and then work with the slice from that point on. So build the slice in `rs_asn1_decode` right after the null checks an then pass the slice to `asn1_decode`. Does that make sense here?

I see what you mean. What do you think about moving the keyword inputs validation (offsets, etc) to rs_asn1_decode, then building the slice?

1. check nulls
2. calculate offset (keyword: absolute_offset, relative_offset)
3. verify we won't read past front/end of buffer
4. read buffer
5. asn1_decode(buffer)

File: rust/src/asn1/parse_rules.rs (hunk 2 above)
Comment (jasonish) [reply rust/src/asn1/parse_rules.rs:94]: usize is variable in size, 4 bytes on 32 bit, 8 bytes on 32 bit.

File: rust/src/asn1/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/asn1/mod.rs:241]: If we do much of the validation in unsafe Rust there isn't a big benefit over doing it in C I think. So I'd prefer doing as little as possible while `unsafe` and then handing of to "safe" Rust for the additional checks

[the remark this answers is not in the document]
Comment (sharksforarms) [reply rust/src/asn1/mod.rs:241]: Ok, I think I can read the entire buffer, then slice it afterwards once the offsets are checked. Will look into this.

**Comment by sharksforarms** [comment]
Continued: https://github.com/OISF/suricata/pull/4965

**Pull request closed without merging**